pkg/git/fetch.go

```diff
@@ -0,0 +1,65 @@
+package git
+
+import (
+	"log"
+
+	"github.com/pkg/errors"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+type Fetcher struct {
+	Logger   *log.Logger
+	Keychain GitKeychain
+}
+
+func (f Fetcher) Fetch(dir, gitURL, gitRevision string) error {
+	resolvedAuth, err := f.Keychain.Resolve(gitURL)
+	if err != nil {
+		return err
+	}
+
+	repo, err := git.PlainInit(dir, false)
+	if err != nil {
+		return errors.Wrap(err, "unable to init git repository")
+	}
+
+	remote, err := repo.CreateRemote(&config.RemoteConfig{
+		Name: "origin",
+		URLs: []string{gitURL},
+	})
+	if err != nil {
+		return errors.Wrap(err, "unable to create remote")
+	}
+
+	opts := &git.FetchOptions{
+		RefSpecs: []config.RefSpec{"refs/*:refs/*"},
+		Auth:     resolvedAuth,
+		Depth:    0,
```

Comment:
Of note, For simplicity in this PR: I removed the depth=1 checkout from our current source-init. It would make sense in a future issue to optimize this checkout.